client-testkit/shared/src/main/scala/org/http4s/client/testkit/WSTestClient.scala

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2014 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s
+package client
+package testkit
+
+import cats.Applicative
+import cats.Foldable
+import cats.MonadThrow
+import cats.effect.Concurrent
+import cats.effect.Resource
+import cats.effect.implicits._
+import cats.effect.std.Queue
+import cats.implicits._
+import fs2.Stream
+import org.http4s.HttpApp
+import org.http4s.client.websocket.WSClient
+import org.http4s.client.websocket.WSConnection
+import org.http4s.client.websocket.WSFrame
+import org.http4s.server.websocket.WebSocketBuilder2
+import org.http4s.websocket.WebSocketCombinedPipe
+import org.http4s.websocket.WebSocketContext
+import org.http4s.websocket.WebSocketFrame
+import org.http4s.websocket.WebSocketSeparatePipe
+
+object WSTestClient {
+
+  def fromHttpWebSocketApp[F[_]: Concurrent](
+      f: WebSocketBuilder2[F] => HttpApp[F]
+  ): F[WSClient[F]] =
```

Comment:
I wonder if we should return `F[(Client[F], WSClient[F])]` or `F[Client[F] with WSClient[F]]`.

Or, offer some kind of "dummy" `WebSocketBuilder2`? (We may have that already and I forgot.)

The question is, what if a user has a `WebSocketBuilder2[F] => HttpApp[F]` application and they want to test the non-websocket routes, how would they do that?

Reply:
Ok, we do have this:
https://github.com/http4s/http4s/blob/bb1668ee6fc3659c1ffc64620e1d025563b50949/server/shared/src/main/scala/org/http4s/server/websocket/WebSocketBuilder2.scala#L190-L191

I think that's good enough!